### ISSUE TITLE
Cargo prints rustc's stderr twice

### ISSUE THREAD

**Issue by Mark-Simulacrum** [rust-lang/cargo#4223]
This is primarily a problem for those who use this output and pipe to editors to jump from error to error. The cause of this is the fact that Cargo uses `exec_with_streaming` inside `exec_json`, and passes [this stderror handler](https://github.com/rust-lang/cargo/blob/master/src/cargo/ops/cargo_rustc/mod.rs#L370) in. That prints out stderr upon receiving each line, which is fine -- we should do this. However, exec_with_streaming also saves this data [here](https://github.com/rust-lang/cargo/blob/abf01e1eddb3145c83f71b469ea7bee37141e5e1/src/cargo/util/process_builder.rs#L165). This output is later printed in "process didn't execute successfully," leading to duplication of the stderr for the compiler.

Arguably, none of these individually are bad, but together they don't lead to optimal behavior. I think the easiest and most straightforward fix here is to add a flag to exec_with_streaming that prevents printing stderr when the command fails and pass it in from cargo_rustc.

cc @alexcrichton 

cc @eddyb -- you asked on IRC about this

**Comment by matklad** [comment]
Hm, this is probably a regression: I remember seeing a similar problem with nightly Cargo recently, while stable printed error only once. I think this happened around after https://github.com/rust-lang/cargo/issues/4029.

**Comment by alexcrichton** [comment]
Ah yeah Cargo shouldn't ever print anything out twice here! @Mark-Simulacrum I think you hit the nail head on. My preferred fix for this would be to stream output but not capture it. That way you get errors and other diagnostics as they come and we shouldn't duplicate anything.

**Comment by Mark-Simulacrum** [comment]
I'll file a PR soon, hopefully.

**Comment by flavius** [comment]
There seems to be a regression on this with `cargo 1.34.0 (6789d8a0a 2019-04-01)` when it's building as part of `cargo test -- --nocapture`

**Comment by ehuss** [comment]
@flavius Can you file a new issue with a reproduction?
